### PR TITLE
Refactor autocomplete

### LINKED ISSUE #669: [AutoComplete] Results are displayed under the radio button
## Actual behaviour

In a form, I use an autocomplete, the results are displayed => OK.

If there a radio input below the autocomplete field, the radio inputs are displayed on the autocomplete results => KO

overflow

## Expected behaviour
The autocomplete results should be displayed on the top of all the form.
## Versions

  * Browser : Chrome
  * `focus` : 0.11.0
  * `focus-components` :  0.7.3-5

### PR DESCRIPTION
**Pull request by Bernardstanislas**
## Autocomplete Select

The Autocomplete has been refactored, and the first component, `AutocompleteSelect` is released now.

image

### Specification

See #688 

### Usage

A new helper is available in the form preset : `this.autocompleteSelectFor('name', {keyResolver, querySearcher});`

### Customization

The options list is customizable, this can be achieve through the `renderOptions` prop.
This prop should return some JSX, and it is bound to the component's context, so the props and the state are directly accessible through `this.props` and `this.state`

### Bug fixes

It fixes #669 #674 #693

### CODE REVIEW

File: src/common/field/index.js [hunk 1]
```diff
@@ -54,16 +54,16 @@ const FieldMixin = {
...
-        const {autocomplete, label, input, select, display} = this;
+        const {autocomplete, autocompleteSelect, label, input, select, display} = this;
         return (
             <div className='mdl-grid' data-domain={domain} data-focus='field' data-mode={isEdit ? 'edit' : 'consult'} data-required={isRequired} data-valid={!error}>
                 {isCustomComponent && this._renderFieldComponent()}
                 {!isCustomComponent && hasLabel && label()}
                 {!isCustomComponent &&
                     <div className ={`${this._getContentGridClassName()}`} data-focus='field-value-container'>
-                        {codeResolver && searcher ? autocomplete() : isEdit ? (values ? select() : input()) : display()}
+                        {codeResolver && searcher ? autocomplete() : keyResolver && querySearcher ? autocompleteSelect() : isEdit ? (values ? select() : input()) : display()}
```
Comment (pierr) [inline src/common/field/index.js]: a bit unreadable...
`isAutoCOmplete`

File: src/components/input/autocomplete-select/consult.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    componentWillReceiveProps({value}) {
+        if (value !== this.props.value) this._callKeyResolver(value);
+    }
+
+    _callKeyResolver(value) {
+        const {keyResolver} = this.props;
+        keyResolver(value).then(label => {
+            this.setState({resolvedLabel: label});
+        }).catch(() => {});
```
Comment (pierr) [inline src/components/input/autocomplete-select/consult.js]: log?

File: src/components/input/autocomplete-select/edit.js [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    _handleDocumentClick = ({target}) => {
+        const {focus, inputValue} = this.state;
+        const {onBadInput} = this.props;
+        if (focus) {
+            const dataset = target ? target.dataset: null;
+            const reactid = dataset ? dataset.reactid : null;
+            const [options, input] = ['options', 'htmlInput'].map(ref => ReactDOM.findDOMNode(this.refs[ref]));
+            const optionsId = options ? options.dataset.reactid : null;
+            const inputId = input ? input.dataset.reactid : null;
```
Comment (pierr) [inline src/components/input/autocomplete-select/edit.js]: I hope `reactid` will not disapear

File: src/components/input/autocomplete-select/edit.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    _handleQueryFocus = () => {
+        this.refs.options.scrollTop = 0;
+        this.setState({active: '', focus: true});
+    };
+
+    _handleQueryKeyDown = (event) => {
+        event.stopPropagation();
+        const {which} = event;
+        const {active, options} = this.state;
+        if (which === 13 && active) this._select(active);
```
Comment (pierr) [inline src/components/input/autocomplete-select/edit.js]: const KEY_NAME = XX

File: src/components/input/autocomplete-select/edit.js (hunk 3 above)
Comment (Bernardstanislas) [inline src/components/input/autocomplete-select/edit.js]: :children_crossing:

File: src/components/input/autocomplete-select/consult.js (hunk 2 above)
Comment (Bernardstanislas) [inline src/components/input/autocomplete-select/consult.js]: Right !

File: src/common/field/index.js (hunk 1 above)
Comment (Bernardstanislas) [inline src/common/field/index.js]: That's intentionnal, I want to keep it dirty to press the urge for the refactor we are working on :)

Comment (pierr) [inline src/common/field/index.js]: :+1:

File: src/components/input/autocomplete-select/edit.js (hunk 4 above)
Comment (Bernardstanislas) [inline src/components/input/autocomplete-select/edit.js]: Right !

File: src/common/field/index.js (hunk 1 above)
Comment (pierr) [inline src/common/field/index.js]: https://twitter.com/JonathanDeMoor/status/676027065171316737

File: src/components/input/autocomplete-select/style.scss [hunk 5]
```diff
@@ -0,0 +1,50 @@
+[data-focus='autocomplete'] {
+    position: relative;
+
+    [data-focussed='true'] {
+        visibility: visible;
+        max-height: 300px;
+    }
+    [data-focussed='false'] {
```
Comment (pierr) [inline src/components/input/autocomplete-select/style.scss]: a bit confusing...

**Pull request merged**